rust_src/src/fonts.rs

```diff
@@ -265,4 +309,50 @@ pub fn font_at_lisp(
     unsafe { font_at(-1, pos, ptr::null_mut(), w.as_mut(), string) }
 }
 
+/// List available fonts matching FONT-SPEC on the current frame.
+/// Optional 2nd argument FRAME specifies the target frame.
+/// Optional 3rd argument NUM, if non-nil, limits the number of returned fonts.
+/// Optional 4th argument PREFER, if non-nil, is a font-spec to
+/// control the order of the returned list.  Fonts are sorted by
+/// how close they are to PREFER.
+#[lisp_fn(min = "1")]
+pub fn list_fonts(
+    font_spec: LispFontSpecRef,
+    frame: LispFrameLiveOrSelected,
+    num: Option<EmacsInt>,
+    prefer: Option<LispFontSpecRef>,
+) -> LispObject {
+    let mut frame: LispFrameRef = frame.into();
+
+    let n = match num {
+        Some(n) if n < 0 => return Qnil,
+        Some(n) => n,
+        None => 0,
+    } as usize;
+
+    let list = unsafe { font_list_entities(frame.as_mut(), font_spec.into()) };
+    match list.into() {
+        Some((car, cdr)) if cdr.is_nil() => match car.as_vector() {
+            Some(vec) if vec.len() == 1 => return list!(vec.get(0)),
+            _ => (),
+        },
+        _ => return Qnil,
+    }
+
+    let vec = match prefer {
+        None => unsafe { font_sort_entities(list, prefer.into(), frame.as_mut(), 0) },
+        Some(_) => vconcat_entity_vectors(list.into()),
+    }
+    .force_vector();
+
+    if n == 0 || n >= vec.len() {
+        fns::append(&mut [list, vec.into(), Qnil])
```

Comment:
Is this doing the same as `list = CALLN (Fappend, vec, Qnil);`? Why is list being included?

Reply:
That was a mixup on my part. List should not be included.